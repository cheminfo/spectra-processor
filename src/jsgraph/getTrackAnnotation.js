import { getNormalizedData } from '../spectra/getNormalizedData.js';

export function getTrackAnnotation(spectra, index, options = {}) {
  const { ids, showSpectrumID = true, sortY = true, limit = 20 } = options;
  let annotations = [];

  let normalized = getNormalizedData(spectra, { ids });

  if (normalized.ids.length === 0) return annotations;
  let line = 0;

  // header containing X coordinate
  annotations.push({
    type: 'line',
    position: [
      { x: `${70}px`, y: `${15 + 15 * line}px` },
      { x: `${85}px`, y: `${15 + 15 * line}px` },
    ],
    strokeWidth: 0.0000001,
    label: {
      size: 16,
      text: `x: ${normalized.x[index].toPrecision(6)}`,
      position: { x: `${130}px`, y: `${20 + 15 * line}px` },
    },
  });
  line++;

  let peaks = [];
  for (let i = 0; i < normalized.ids.length; i++) {
    peaks.push({
      id: normalized.ids[i],
      meta: normalized.meta[i],
      y: normalized.matrix[i][index],
    });
  }

  if (sortY) {
    peaks = peaks.sort((a, b) => b.y - a.y);
  }
  if (limit) {
    peaks = peaks.slice(0, limit);
  }

  for (let { id, meta, y } of peaks) {
    annotations.push({
      type: 'line',
      position: [
        { x: `${70}px`, y: `${15 + 15 * line}px` },
        { x: `${85}px`, y: `${15 + 15 * line}px` },
      ],
      strokeColor: meta.color,
      strokeWidth: 2,
      label: {
        text: `${y.toPrecision(4)}${showSpectrumID ? ` - ${id}` : ''}`,
        position: { x: `${90}px`, y: `${20 + 15 * line}px` },
      },
    });
    line++;
  }

  return annotations;
}
