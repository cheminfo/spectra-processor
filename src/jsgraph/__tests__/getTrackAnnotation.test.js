import { describe, it, expect } from 'vitest';

import { getTrackAnnotation } from '../getTrackAnnotation';

let spectra = [
  {
    normalized: {
      x: [10, 20, 30],
      y: [1, 2, 3],
    },
    meta: {
      color: 'red',
    },
    id: 1,
  },
  {
    normalized: {
      x: [10, 20, 30],
      y: [2, 3, 4],
    },
    meta: {
      color: 'green',
    },
    id: 2,
  },
];

describe('getTrackAnnotation', () => {
  it('check jsgraph annotations', () => {
    let result = getTrackAnnotation(spectra, 1);
    expect(result).toMatchSnapshot();
  });
});
