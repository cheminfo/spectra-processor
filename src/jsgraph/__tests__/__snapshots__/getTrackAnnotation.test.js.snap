// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTrackAnnotation check jsgraph annotations 1`] = `
[
  {
    "label": {
      "position": {
        "x": "130px",
        "y": "20px",
      },
      "size": 16,
      "text": "x: 20.0000",
    },
    "position": [
      {
        "x": "70px",
        "y": "15px",
      },
      {
        "x": "85px",
        "y": "15px",
      },
    ],
    "strokeWidth": 1e-7,
    "type": "line",
  },
  {
    "label": {
      "position": {
        "x": "90px",
        "y": "35px",
      },
      "text": "3.000 - 2",
    },
    "position": [
      {
        "x": "70px",
        "y": "30px",
      },
      {
        "x": "85px",
        "y": "30px",
      },
    ],
    "strokeColor": "green",
    "strokeWidth": 2,
    "type": "line",
  },
  {
    "label": {
      "position": {
        "x": "90px",
        "y": "50px",
      },
      "text": "2.000 - 1",
    },
    "position": [
      {
        "x": "70px",
        "y": "45px",
      },
      {
        "x": "85px",
        "y": "45px",
      },
    ],
    "strokeColor": "red",
    "strokeWidth": 2,
    "type": "line",
  },
]
`;
